rescript/ncbi.py

```diff
@@ -0,0 +1,191 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import time
+import warnings
+
+import requests
+from xmltodict import parse
+from pandas import DataFrame
+from q2_types.feature_data import DNAIterator
+from skbio import DNA
+from qiime2 import Metadata
+from collections import OrderedDict
+
+_default_ranks = [
+    'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'
+]
+_default_allowed_ranks = [
+    'domain', 'superkingdom', 'kingdom', 'subkingdom', 'superphylum', 'phylum',
+    'subphylum', 'infraphylum', 'superclass', 'class', 'subclass',
+    'infraclass', 'superorder', 'order', 'suborder', 'superfamily', 'family',
+    'subfamily', 'genus', 'species'
+]
+
+
+def get_ncbi_data(
+        query: str = None, accession_ids: Metadata = None,
+        ranks: list = None, allowed_ranks: list = None,
+        disable_rank_propagation: bool = False,
+        entrez_delay: float = 0.334) -> (DNAIterator, DataFrame):
+    if query is None and accession_ids is None:
+        raise ValueError('Query or accession_ids must be supplied')
+    if ranks is None:
+        ranks = _default_ranks
+    if allowed_ranks is None:
```

Comment:
I second @nbokulich's idea.

Reply:
Ok, but to do that we have to come up with an exhaustive list of the trusted `allowed_ranks` across all the clades of the tree of life, and we would have to update it whenever someone thinks of a new one.

For instance, the current list does not include "Divisions", "Tribes", and "Varieties".

Is it possible to get an exhaustive list?

Reply:
if users really need some newfangled rank name, we can fix (or they can put in a PR)

I don't think there will be an infinite variety of rank names... the most common rank names will do for us I think.

I'd like to avoid the alternative, in which users run into all sorts of issues by specifying their own...

Reply:
That's a fair point. Do either of you have a slightly more exhaustive list than the current allowed_ranks? This conversation started because we wanted to make the system cater for more than microbes.

Reply:
Please select which of these you would like in the canonical allowed ranks:

Domain
Subdomain
Hyperkingdom
Superkingdom
Kingdom
Subkingdom
Infrakingdom
Parvkingdom
Superphylum
Phylum
Subphylum
Infraphylum
Microphylum
Superclass
Class
Subclass
Infraclass
Parvclass
Superdivision
Division
Subdivision
Infradivision
Superlegion
Legion
Sublegion
Infralegion
Supercohort
Cohort
Subcohort
Infracohort
Superorder
Gigaorder
Magnorder
Grandorder
Mirorder
SeriesFish
Order
Nanorder
Hypoorder
Suborder
Infraorder
Parvorder
Section
Subsection
Gigafamily
Megafamily
Grandfamily
Hyperfamily
Superfamily
Epifamily
SeriesLepidoptera
GroupLepidoptera
Family
Subfamily
Infrafamily
Supertribe
Tribe
Subtribe
Infratribe
Genus
Subgenus
SeriesBotany
SubseriesBotany
SpeciesGroup
SpeciesSubgroup
Species
Subspecies
Varietas
Subvarietas
Forma
Subforma
Cultivar
Breed
Individual
Unspecified

With thanks to https://docs.rs/taxonomy/0.3.3/taxonomy/enum.TaxRank.html

Reply:
ok, I've hard coded the allowed ranks and made appropriate changes to the rank propagation code